ocaml/xapi/xapi_pci_helpers.ml

```diff
@@ -172,3 +173,62 @@ let get_host_pcis () =
 let igd_is_whitelisted ~__context pci =
   let vendor_id = Db.PCI.get_vendor_id ~__context ~self:pci in
   List.mem vendor_id !Xapi_globs.igd_passthru_vendor_whitelist
+
+let is_pci_hidden_cmdline ~__context ~self =
+  let cmdline =
+    match Unixext.read_lines ~path:"/proc/cmdline" with
+    | [x] ->
+        x
+    | _ ->
+        failwith "Unable to read cmdline"
+  in
+  let device = Db.PCI.get_pci_id ~__context ~self in
+  let elems = String.split_on_char ' ' cmdline in
+
+  let pciback =
+    List.find_opt
+      (fun s -> String.starts_with ~prefix:"xen-pciback.hide=" s)
+      elems
+  in
+  (* Look for the device id in the list of hidden devices
+   * pciback looks like: "xen-pciback.hide==(<id1>)(<id2>)..." *)
+  let contains str substr = Astring.String.is_infix ~affix:substr str in
+  match pciback with None -> false | Some value -> contains value device
+
+let determine_dom0_access_status ~__context ~self =
+  (* Current hidden status *)
+  let is_hidden_cmdline = is_pci_hidden_cmdline ~__context ~self in
+  (* Hidden status after reboot *)
+  let is_hidden = Pciops.is_pci_hidden ~__context self in
+  let new_access =
+    match (is_hidden_cmdline, is_hidden) with
+    | true, true ->
+        `disabled
+    | false, true ->
+        `disable_on_reboot
+    | false, false ->
+        `enabled
+    | true, false ->
+        `enable_on_reboot
+  in
+
+  (* Keep up to date deprecated PGPU DB field, to be removed eventually. *)
+  let expr = Printf.sprintf {|field "PCI"="%s"|} (Ref.string_of self) in
+  let pgpus = Db.PGPU.get_all_records_where ~__context ~expr in
+  List.iter
+    (fun (pgpu_ref, _) ->
+      Db.PGPU.set_dom0_access ~__context ~self:pgpu_ref ~value:new_access
+    )
+    pgpus ;
```

Comment:
This should be moved to `update_dom0_access`, the name makes it clear side-effects are expected

Reply:
I disagree, putting it here allows to check and update the deprecated field when the `get_dom0_access_status` is called as well and so this will reflect the changes done outside XAPI if a user has called the xen cmdline script manually.

Reply:
> putting it here allows to check and update the deprecated field 

This is what I don't like, this is not the place to update the database field, the place to do it is when the state changes:
- User has called the API to change the state
- Host reboots

Please move the update code to its own function, call it from update_dom0_access; and add a new function that updates all existing pgpu objects so it can be called on xapi startup.

Reply:
Done: The PGPU db field update code is now in `update_dom0_access` so only when a user call the API to change the `dom0_access_status`
The sync as reboot was already working so nothing to do here.

It means that the only desync is: if a user un/hides a PCI without using the API (which is not supported as you said) the state `enable/disable_on_reboot` is not displayed for the PGPU, but everything will be synced after the reboot.